### PR TITLE
Python3 fix

### PR DESCRIPTION
**Pull request by nyov**
This makes it work for me (Python 3.5).

### CODE REVIEW

**Comment by kalouantonis** [comment]
Thanks for the contribution!

CI seems to pass just fine for python 2.7, so that's good.

Could you add a line in the `.travis.yml` CI file to include python upwards of 3.4 too?

**Comment by nyov** [comment]
I can, but the testcases aren't py3 compatible from what I can tell.

**Comment by nyov** [comment]
I fixed the tests to complete on PY3, however no guarantee that they verify what they're supposed to.

I believe there is at least one serious issue though, from seeing some of the unfixed tests HANG without completing for forever (killed after 10mins by travis): https://travis-ci.org/johan-sports/pyheatshrink/builds/412733650
That shouldn't ever happen, even if I feed crap into the `EncodedFile` routine (as the unfixed test `test_read1 (tests.test_streams.EncodedFileTest)` apparently does on PY3), because that signifies a potential DOS security bug.

File: tests/test_streams.py [hunk 1]
```diff
@@ -98,17 +104,21 @@ def test_with_large_files(self):
...
 
             with EncodedFile(TEST_FILENAME, mode='wb') as fp:
                 fp.write(contents)
 
             with EncodedFile(TEST_FILENAME) as fp:
                 # Read small buffer sizes
-                read_func = functools.partial(fp.read, 512)
-                read_str = ''.join([s for s in iter(read_func, '')])
+                read_str = fp.read()
+                # FIXME!
```
Comment (kalouantonis) [inline tests/test_streams.py]: How come you marked this one as FIXME?

**Comment by kalouantonis** [comment]
A quick glance at the code didn't bring up anything immediately obvious as to why the tests hang like that.

The CI tests for your latest changes on this branch don't seem to cause failures however: https://travis-ci.org/johan-sports/pyheatshrink/builds/412739550. Perhaps the Cython code needs updating somewhere? I took reasonable care to keep it agnostic.

I have to note that I can't accept this PR until the quality of the python 3 implementation matches that of the python 2 implementation.

**Comment by nyov** [comment]
Yes, the fixed testcases pass and don't hang anymore. I'm just mentioning that you can feed data into `EncodedFile` that will make it hang forever.

> I can't accept this PR until the quality of the python 3 implementation matches that of the python 2 implementation.

You misunderstand. I'm not doing a py3 implementation or port here; I merely contributed a fix for a bad reference to PY2 `__builtin__` - which doesn't exist in PY3.
And then I contributed some possible fixes for the test cases as well in response your Travis CI request.
The "Python 3 -- Experimental" support is already listed in the Readme - I didn't add that. And sadly I don't have more time to debug this codebase (that I'm not using). Feel free to disregard this PR.

**Comment by kalouantonis** [comment]
Ah, I misunderstood what you were saying. Your effort is highly appreciated and you went above and beyond what you had to.

I'll look into what's going on with `read1` and if its an exploitable issue on the current master or if this PR introduces it, which I'll look into fixing either way when I get some time.

Your changes are really nice btw, since python 2 is practically end of life.

**Comment by nyov** [comment]
I fixed the "FIXME" in the testcase (which is where it hang while I rewrote tests), sorry I forgot about that. It seems to work now, though.

edit: And another update for better style. I hope the testcase changes are now straightforward.

**Comment by chetbox** [comment]
Is this ready to merge? It would be very helpful to have this merged and available through `pip`.

Edit: For others looking to use this you can install it using `pip install git+git://github.com/nyov/pyheatshrink.git@3294ace96113e22a788573357c14120bc95ab55a#egg=heatshrink`.

**Comment by nyov** [comment]
OH, look what I wrote back in ...2018. I did not even remember this one, haha.  

Well, I don't see why it shouldn't be merge-able. Some integration testing by others (like you) should be done though, since the unittests were changed.

This means for Python3 the code was changed to handle `bytes` instead of `unicode` strings -- which you'll see happening everywhere else in Py3 libs as well, such as zip-file and other compressors, as it doesn't make sense to only be able to handle text-strings there -- and once tests confirm that it works as expected, people could report back.

Maybe someone will actually merge it then.

**Comment by chetbox** [comment]
I've tested compressing, decompressing and verified the input matches output. I can confirm this is working on Python 3.6 on the `python:3.6` Docker image.

@kalouantonis can we merge this?

**Comment by jsiverskog** [comment]
@kalouantonis should this be merged?

**Comment by dsnipe** [comment]
@jsiverskog I have access to this repository and can merge if @kalouantonis won't answer soon.

**Comment by yawor** [comment]
@dsnipe would it also be possible to bump the version number and publish this to pypi?

**Comment by dsnipe** [comment]
@yawor, unfortunately, I don't have access to PyPi. @kalouantonis has it.
I can grant access to the repository to a new maintainer, or better - somebody could fork it and maintain it there.
In the company, we don't use Python anymore. I can put a link to a fork in README. 
@kalouantonis maybe you want to take it? 
Sorry for inconvinient, guys!

**Comment by dsnipe** [comment]
I'm going to open a new issue and copy this message there.

**Pull request merged**